File: ruby/0067.rb
def invalid_xy?(x, y)
  (x < 0 or x > 11 or y < 0 or y > 11)
end

def remove_island(x, y)
  [[1, 0], [-1, 0], [0, 1], [0, -1]].each do |dx, dy|
    next if invalid_xy?(x + dx, y + dy)
    if @map[y + dy][x + dx] == 1
      @map[y + dy][x + dx] = 0
      remove_island(x + dx, y + dy)
    end
  end
end

loop do
  @map = []
  while s = gets do
    tmp = s.chomp.split('').map(&:to_i)
    break unless tmp.count > 0
    @map << tmp
  end
  break unless @map.size == 12

  count = 0
  0.upto(11) do |x|
    0.upto(11) do |y|
      if @map[y][x] == 1
        count += 1
        remove_island(x, y)
      end
    end
  end
  puts count
end

File: python/0067.py
def get_map():
  map = []
  while True:
    try:
      tmp = list(raw_input())
      if len(tmp) != 12: break
      map.append(tmp)
    except:
      break
  return map

def remove_island(x, y, map):
  map[x][y] = 0
  move = [[1, 0], [0, 1], [-1, 0], [0, -1]]
  for i, j in move:
    if 0 <= x + i <= 11 and 0 <= y + j <= 11 and map[x + i][y + j] == '1':
      map = remove_island(x + i, y + j, map)
  return map[:]

while True:
  map = get_map()
  if len(map) != 12: break

  count = 0
  for x in range(12):
    for y in range(12):
      if map[x][y] == '1':
        count += 1
        map = remove_island(x, y, map)
  print count
